rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/deduplication/TbMsgDeduplicationNode.java

```diff
@@ -0,0 +1,249 @@
+import org.thingsboard.rule.engine.api.TbNode;
+import org.thingsboard.rule.engine.api.TbNodeConfiguration;
+import org.thingsboard.rule.engine.api.TbNodeException;
+import org.thingsboard.rule.engine.api.TbRelationTypes;
+import org.thingsboard.rule.engine.api.util.TbNodeUtils;
+import org.thingsboard.server.common.data.id.EntityId;
+import org.thingsboard.server.common.data.plugin.ComponentType;
+import org.thingsboard.server.common.msg.TbMsg;
+import org.thingsboard.server.common.msg.TbMsgMetaData;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+@RuleNode(
+        type = ComponentType.ACTION,
+        name = "deduplication",
+        configClazz = TbMsgDeduplicationNodeConfiguration.class,
+        nodeDescription = "Deduplicate messages for a configurable period based on a specified deduplication strategy.",
+        nodeDetails = "Rule node allows you to select one of the following strategy to deduplicate messages: <br></br>" +
+                "<b>FIRST</b> - return first message that arrived during deduplication period.<br></br>" +
+                "<b>LAST</b> - return last message that arrived during deduplication period.<br></br>" +
+                "<b>ALL</b> - return all messages as a single JSON array message. Where each element represents object with <b>msg</b> and <b>metadata</b> inner properties.<br></br>",
+        icon = "content_copy",
+        uiResources = {"static/rulenode/rulenode-core-config.js"},
+        configDirective = "tbActionNodeMsgDeduplicationConfig"
+)
+@Slf4j
+public class TbMsgDeduplicationNode implements TbNode {
+
+    private static final String TB_MSG_DEDUPLICATION_TIMEOUT_MSG = "TbMsgDeduplicationNodeMsg";
+    private static final int TB_MSG_DEDUPLICATION_TIMEOUT = 5000;
+    public static final int TB_MSG_DEDUPLICATION_RETRY_DELAY = 10;
+
+    private TbMsgDeduplicationNodeConfiguration config;
+
+    private Map<EntityId, List<TbMsg>> deduplicationMap;
+    private long deduplicationInterval;
+    private long lastScheduledTs;
+    private DeduplicationId deduplicationId;
+
+
+    @Override
+    public void init(TbContext ctx, TbNodeConfiguration configuration) throws TbNodeException {
+        this.config = TbNodeUtils.convert(configuration, TbMsgDeduplicationNodeConfiguration.class);
+        this.deduplicationInterval = TimeUnit.SECONDS.toMillis(config.getInterval());
+        this.deduplicationId = config.getId();
+        this.deduplicationMap = new HashMap<>();
+        scheduleTickMsg(ctx);
+    }
+
+    @Override
+    public void onMsg(TbContext ctx, TbMsg msg) throws ExecutionException, InterruptedException, TbNodeException {
+        if (TB_MSG_DEDUPLICATION_TIMEOUT_MSG.equals(msg.getType())) {
+            processDeduplication(ctx);
+            scheduleTickMsg(ctx);
```

Comment:
I believe you should scheduleTickMsg in a "finally" block. Otherwise any runtime error in the "processDeduplication" will stop future processing attempts